src/simplemmap.c

```diff
@@ -365,8 +365,114 @@ static void InitMmapRealClass(DllInfo *dll)
  * Constructor
  */
 
-#ifdef Win32
-# error "I'm sure this needs adjusting for Windows, so punt for now."
+//#define DEBUG_PRINT(x) REprintf(x);
+#define DEBUG_PRINT(x) do { } while (0)
+
+#define MMAP_FILE_WARNING_OR_ERROR(str, ...) do {   \
+    if (warn) {                 \
+        warning(str, __VA_ARGS__);          \
+        return NULL;                \
+    }                       \
+    else error(str, __VA_ARGS__);           \
+    } while (0)
+
+#ifdef _WIN32
+
+#include <windows.h>
+
+static void mmap_finalize(SEXP eptr)
+{
+    DEBUG_PRINT("finalizing ... ");
+    void *p = R_ExternalPtrAddr(eptr);
+    R_SetExternalPtrAddr(eptr, NULL);
+    if (p != NULL) {
+        UnmapViewOfFile(p); /* don't check for errors */
+        R_SetExternalPtrAddr(eptr, NULL);
+    }
+    DEBUG_PRINT("done\n");
+}
+
+static LPSTR GetErrorMessage()
+{
+    static CHAR buf[256];
+    FormatMessageA(
+        FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS,
+        NULL,
+        GetLastError(),
+        MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
+        buf,
+        sizeof(buf),
+        NULL
+    );
+    size_t i = strlen(buf);
+    if (i > 0 && buf[i - 1] == '\n')
+        buf[--i] = '\0';
+    if (i > 0 && buf[i - 1] == '\r')
+        buf[--i] = '\0';
+    if (i > 0 && buf[i - 1] == '.')
+        buf[--i] = '\0';
+    return buf;
+}
+
+static SEXP mmap_file(SEXP file, int type, Rboolean ptrOK, Rboolean wrtOK,
+                      Rboolean serOK, Rboolean warn)
+{
+    const char *efn = R_ExpandFileName(translateChar(STRING_ELT(file, 0)));
+
+    /* Open file */
+    HANDLE hFile = CreateFileA(
```

Comment:
I should use `filenameToWchar` and `CreateFileW` here, but `filenameToWchar` is only available in base R. I will try to extract a version of this function that doesn't need to guess the encoding (these macros are also only available internally) by converting the filename to UTF-8 first.